src/base/utils/tar.cpp

```diff
@@ -0,0 +1,298 @@
+
+#include "base/logger.h"
+#include "base/utils/bytearray.h"
+#include "base/utils/fs.h"
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    const int TAR_FILENAME_OFFSET = 0;
+    const int TAR_FILESIZE_OFFSET = 124;
+    const int TAR_CHECKSUM_OFFSET = 148;
+    const int TAR_FILETYPE_OFFSET = 156;
+    const int TAR_FILENAME_PREFIX_OFFSET = 345;
+
+    const int TAR_FILENAME_SIZE = 100;
+    const int TAR_FILESIZE_SIZE = 12;
+    const int TAR_CHECKSUM_SIZE = 8;
+    const int TAR_FILENAME_PREFIX_SIZE = 155;
+
+    class TarEntry
+    {
+    public:
+        enum Types : unsigned char
+        {
+            REG_FILE_NULL = '\0',
+            REG_FILE_ASCII = '0',
+            // We don't care about the other file types
+
+            INVALID = std::numeric_limits<unsigned char>::max()
+        };
+
+        TarEntry() = default;
+
+        explicit TarEntry(const QByteArray &header)
+            : m_header(header)
+        {
+            if (m_header.size() != TAR_HEADER_SIZE)
+                return;
+
+            // Invalid header: The first header field represents the filename.
+            // If the 1st byte is null then we can't extract a name. So the whole header becomes invalid to us.
+            m_valid = !m_header.startsWith('\0');
+            verifyAndCheckNull();
+        }
+
+        QString name() const
+        {
+            if (!isValid())
+                return {};
+
+            const QByteArray filename = Utils::ByteArray::midView(m_header, TAR_FILENAME_OFFSET, TAR_FILENAME_SIZE);
+            const int filenameLen = strlen(filename);
+            if (filenameLen == 0)
+                return {};
+
+            const QByteArray filenamePrefix = Utils::ByteArray::midView(m_header, TAR_FILENAME_PREFIX_OFFSET, TAR_FILENAME_PREFIX_SIZE);
+            const int filenamePrefixLen = strlen(filenamePrefix);
+
+            const QString finalFilename = (filenamePrefixLen > 0) ?
+                                        QString::fromUtf8(filenamePrefix.constData(), filenamePrefixLen)
+                                              + '/' + QString::fromUtf8(filename.constData(), filenameLen)
+                                      : QString::fromUtf8(filename.constData(), filenameLen);
+
+            return Utils::Fs::toValidFileSystemName(finalFilename, true);
+        }
+
+        bool isNull() const { return m_null; }
+        bool isValid() const { return m_valid; }
+        bool isVerified() const { return m_verified; }
+        Types type() const { return isValid() ? static_cast<Types>(m_header.at(TAR_FILETYPE_OFFSET)) : INVALID; }
+        int size() const { return isValid() ? octalToDec(Utils::ByteArray::midView(m_header, TAR_FILESIZE_OFFSET, TAR_FILESIZE_SIZE)) : 0; }
+
+    private:
+        int octalToDec(const QByteArray &data) const
+        {
+            // only the ASCII characters from '0' to '7' are valid
+            int i = 0;
+            for (; i < data.size(); ++i) { // skip leading invalid bytes
```

Comment:
Looks suspicious. IMO, you provide some kind of undefined behaviour.
You should fail on really invalid characters.
If spaces at beginning are acceptable (i.e. "   354" is valid octal number) you should skip spaces explicitly. But, e.g., "gdi754" is invalid octal number and it should cause an exception.

Reply:
Sorry for the lack of activity. I'll probably push something the next days based on the various comments. On this particular one:
IIRC the specs calls that the field is an octal number padded with leading zeroes (all ASCII chars). So I am going to interpret it this ways:
```c++
int octalToDec(const QByteArray &data) const
{
int fileSize = 0;

  for (int i=0; i < data.size(); ++i) {
    char byte = data.at(i);
      if (byte < '0' || byte > '7') // skip trailing invalid bytes
        break;

    fileSize = (8 * fileSize) + (byte - '0');
  }

return fileSize;
}
```

Reply:
>IIRC the specs calls that the field is an octal number padded with leading zeroes (all ASCII chars).

If it's true, then your current algorithm is correct, but it can be used only internally by tar parser (so we should never make it shareable). Or you can split it into two functions: first (can be shareable util) parses valid octal notation (and breaks at any invalid character), second (private) skips leading invalid characters and apply the first one on the rest.